### PR TITLE
Koin dsl

### PR DESCRIPTION
**Pull request by abelgardep**
Koin now offer a new kind of DSL keyword that allow you to target a class constructor directly, and avoid to to have type your definition within a lambda expression.

Use [koin constructor dsl](https://insert-koin.io/docs/reference/koin-core/dsl-update/) to inject view models, repositories, data sources etc.. Its a little bit cleaner now, and no need to introduce `get() get() get()` over the whole module.

### CODE REVIEW

**Comment by abelgardep** [comment]
I think that Koin DSL makes it easier to define the dependencies. It removes some verbosity that we have in the standard way.

If you consider it useful, I can fix the conflicts. Otherwise, close the pull request 👍 

@jesmrec @Aitorbp

**Comment by jesmrec** [comment]
We are still interested, will not be closed. As usual, time reduced. Thanks a lot.

**Comment by abelgardep** [comment]
I'm not able to fix the conflicts because I don't have write access to the repository anymore 😢 

I could try to cherry-pick and create a new PR from my fork though. Otherwise, you will need to fix the conflicts yourselves.

So up to you 👍

**Comment by jesmrec** [comment]
We'll try to fix first. Thanks again for your engagement.

**Comment by michaelstingl** [comment]
> I don't have write access to the repository anymore 😢

Should be fixed 👍

**Comment by abelgardep** [comment]
Rebased.

Little consideration: We can't inject over 10 dependencies using koin dsl. For example, that's why I kept `TransfersViewModel` the way it was.

File: owncloudApp/src/main/java/com/owncloud/android/dependecyinjection/LocalDataSourceModule.kt [hunk 1]
```diff
@@ -63,16 +66,16 @@ val localDataSourceModule = module {
     single { OwncloudDatabase.getDatabase(androidContext()).transferDao() }
     single { OwncloudDatabase.getDatabase(androidContext()).spacesDao() }
 
-    single<SharedPreferencesProvider> { OCSharedPreferencesProvider(get()) }
+    singleOf(::OCSharedPreferencesProvider) bind SharedPreferencesProvider::class
     single<LocalStorageProvider> { ScopedStorageProvider(dataFolder, androidContext()) }
 
     factory<LocalAppRegistryDataSource> { OCLocalAppRegistryDataSource(get()) }
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/dependecyinjection/LocalDataSourceModule.kt:72]: Can't this be converted to the new syntax as well? We just have a `get()`, or am I missing something?

File: owncloudApp/src/main/java/com/owncloud/android/dependecyinjection/LocalDataSourceModule.kt [hunk 2]
```diff
@@ -63,16 +66,16 @@ val localDataSourceModule = module {
...
-    factory<FolderBackupLocalDataSource> { OCFolderBackupLocalDataSource(get()) }
-    factory<LocalTransferDataSource> { OCLocalTransferDataSource(get()) }
-    factory<LocalSpacesDataSource> { OCLocalSpacesDataSource(get()) }
+    factoryOf(::OCFolderBackupLocalDataSource) bind FolderBackupLocalDataSource::class
+    factoryOf(::OCLocalCapabilitiesDataSource) bind LocalCapabilitiesDataSource::class
+    factoryOf(::OCLocalFileDataSource) bind LocalFileDataSource::class
+    factoryOf(::OCLocalShareDataSource) bind LocalShareDataSource::class
+    factoryOf(::OCLocalTransferDataSource) bind LocalTransferDataSource::class
+    factoryOf(::OCLocalUserDataSource) bind LocalUserDataSource::class
+    factoryOf(::OCLocalSpacesDataSource) bind LocalSpacesDataSource::class
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/dependecyinjection/LocalDataSourceModule.kt:80]: I think you tried to order them alphabetically, this one is not in its corresponding place then 😃

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some tiny comments here @abelgardep 🍻

**Comment by abelgardep** [comment]
Changes applied and rebased @JuancaG05 🍻

**Review by JuancaG05 — APPROVED** [review]
LGTM! Thanks for your contribution @abelgardep! 🚀

**Pull request merged**